### PR TITLE
Using 'shared' variables without synchronization

### LINKED ISSUE #1676: using 'shared' variables without synchronization
Some of our intermittent test failures might be caused by not using synchronization(locks/atomic operations) while using 'shared' variables for example:

SlashingMisbehavingValidator uses
private shared bool* reveal_preimage;
from multiple threads without any syncronization

Some other affected tests:
Byzantine.d
EnrollmentManager.d

### PR DESCRIPTION
**Pull request by ferencdg**
fixes: #1676

### CODE REVIEW

File: source/agora/test/SlashingMisbehavingValidator.d
Comment (ferencdg) [inline source/agora/test/SlashingMisbehavingValidator.d]: will put back the 'shared' here, even though there is no compiler enforcement

**Comment by Geod24** [comment]
I was more hoping we could remove the usage of a shared UTXOSet instead of endorsing it.

**Comment by ferencdg** [comment]
> I was more hoping we could remove the usage of a shared UTXOSet instead of endorsing it.

if we want to use the UTXOSet:

First, I was thinking about creating a new method on the test clients that takes a copy of the exisitng UTXOSet and returns it... However UTXOSet uses an in memory database, so not sure how easy would that be to 'copy'. 

I could
1. query the in-memory DB and return a copy all UTXOs and write methods that search on that returned object 
2. look into how to make a copy of the in-memory DB, so I can copy UTXOSet
3. make a new derived class of UTXOSet that I can easily copy(doesn't have in memory-db)
All of the above would have taken me more time, then the solution I chose.

At the end I just synced the UTXOSet, let me know which version you prefer

---------------------------
if we don't want to use the UTXOSet

We could maybe use the header to know who was slashed, but the original test tests for more than just who was slashed...

**Comment by ferencdg** [comment]
also rebased, let me know if any changes are needed

**Comment by Geod24** [comment]
Can you rebase ? Let's go with the solution we have now, and if we end up extending the test interface, this will benefit from it.

**Comment by ferencdg** [comment]
rebased

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
The timeout seems consistent.

File: doc/Style.md
Comment (ferencdg) [inline doc/Style.md:352]: makeTestNetwork can no longer figure out the default type after signature change(seems Dlang compiler problem)

File: source/agora/test/SlashingMisbehavingValidator.d
Comment (ferencdg) [inline source/agora/test/SlashingMisbehavingValidator.d:69]: why '__gshared'? Using shared would have required us to define all methods in UTXOSet as 'shared'

**Comment by ferencdg** [comment]
> The timeout seems consistent.

I regret ever trying to fix this test... Should work now, but the PR needs a second review.

**Comment by ferencdg** [comment]
🟢 let me know if anything else is needed

File: doc/Style.md
Comment (Geod24) [reply doc/Style.md:352]: Having to modify all tests is a non-starter. I'll take a look at the failure soon. The rest looks good.

**Comment by Geod24** [comment]
Just needs a rebase and let's merge this

**Comment by ferencdg** [comment]
rebased

**Review by Geod24 — APPROVED** [review]

**Pull request merged**
